# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

##读取数据

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

评估数据整洁度，先随机选择10条数据观测

In [4]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
466079    576303    72349B      SET/6 PURPLE BUTTERFLY T-LIGHTS         4   
333219    566189    47599B                     BLUE PARTY BAGS          6   
464045    576076    84596B   SMALL DOLLY MIX DESIGN ORANGE BOWL         5   
96096     544476     21976       PACK OF 60 MUSHROOM CAKE CASES         2   
96105     544476     22659              LUNCH BOX I LOVE LONDON         2   
126383    547074     22996    TRAVEL CARD WALLET VINTAGE TICKET        24   
419798    572859     22766                  PHOTO FRAME CORNICE         1   
182752    552572     23199                     JUMBO BAG APPLES       100   
195211    553707     23203            JUMBO BAG DOILEY PATTERNS        10   
112333    545895     22182  CAKE STAND VICTORIAN FILIGREE SMALL         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
466079  11/14/2011 14:51       2.10     15532.0  United Kingdom  
333219    9/9/2011 13:19       2.10     17800.0  United Kingdom  
464045  11/13/2011 16:18       0.42     14382.0  United Kingdom  
96096     2/21/2011 9:58       0.55     14796.0  United Kingdom  
96105     2/21/2011 9:58       1.95     14796.0  United Kingdom  
126383   3/20/2011 14:35       0.42     13102.0  United Kingdom  
419798  10/26/2011 12:39       2.95     16418.0  United Kingdom  
182752   5/10/2011 12:11       1.79     14680.0  United Kingdom  
195211   5/18/2011 15:26       2.08     16843.0  United Kingdom  
112333    3/7/2011 17:14       4.13         NaN  United Kingdom

从抽样的10行数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

评估数据干净程度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而description,CustomerID变量存在缺失值。
此外，InvoiceDate的数据类型应为日期，CustomerID的数据类型应为字符串，应当进行数据格式转换。

脏数据可能存在的问题，包括：确实数据，重复数据，不一致数据，无效数据或错误数据

接下来筛选出Unitprice为负数的观察值

In [7]:
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，UnitPrice为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除

## 清理数据
根据前面评估部分得到的结论，我们需要进行的数据清理包括：\
-把InvoiceDate变量的数据类型转换为日期时间\
-把CustomerID变量的数据类型转换为字符串\
-把Description变量缺失的观察值删除\
-把Country变量值"USA"替换为"United States"\
-把Country变量值"UK","U.K"替换为"United Kingdom"\
-把Quantity变量值为负数的观察值删除\
-把UnitPrice变量值为负数的观察值删除

In [10]:
cleaned_data=original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

-把InvoiceDate变量的数据类型转换为日期时间

In [11]:
cleaned_data["InvoiceDate"]=pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

-把CustomerID变量的数据类型转换为字符串

In [14]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的".0"删除

In [16]:
cleaned_data["CustomerID"]=cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

-把Description变量缺失的观察值删除,并查看删除后该列空缺值个数和：

In [18]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [19]:
cleaned_data["Description"].isnull().sum()

0

-把Country变量值"USA"替换为"United States",并检查替换后的"USA变量值个数"；

In [22]:
cleaned_data["Country"]=cleaned_data["Country"].replace({"USA":"United States"})

In [23]:
len(cleaned_data[cleaned_data["Country"]=="USA"])

0

-把Country变量值"UK","U.K"替换为"United Kingdom"

In [29]:
cleaned_data["Country"]=cleaned_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})
print(len(cleaned_data[cleaned_data["Country"]=="UK"]))
print(len(cleaned_data[cleaned_data["Country"]=="U.K."]))

0
0


-把Quantity变量值为负数的观察值删除,并检查替换后Quantity变量值为负数的个数：

In [31]:
cleaned_data=cleaned_data[cleaned_data["Quantity"]>=0]
len(cleaned_data[cleaned_data["Quantity"]<0])

0

In [32]:
cleaned_data=cleaned_data[cleaned_data["UnitPrice"]>=0]
len(cleaned_data[cleaned_data["UnitPrice"]<0])

0

## 保存清理后的数据

完在数据清理后，把干净整齐的数据保存到新的文件里，文件名为：e_commerce_cleaned.csv


In [34]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)

In [36]:
pd.read_csv("e_commerce_cleaned.csv").head(20)

C:\Users\40375\AppData\Local\Temp\ipykernel_11924\2220786711.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head(20)


InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15    536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16    536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17    536367     21754             HOME BUILDING BLOCK WORD         3   
18    536367     21755             LOVE BUILDING BLOCK WORD         3   
19    536367     21777          RECIPE BOX WITH METAL HEART         4   

            InvoiceDate  UnitPrice CustomerID         Country  
0   2010-12-01 08:26:00       2.55      17850  United Kingdom  
1   2010-12-01 08:26:00       3.39      17850  United Kingdom  
2   2010-12-01 08:26:00       2.75      17850  United Kingdom  
3   2010-12-01 08:26:00       3.39      17850  United Kingdom  
4   2010-12-01 08:26:00       3.39      17850  United Kingdom  
5   2010-12-01 08:26:00       7.65      17850  United Kingdom  
6   2010-12-01 08:26:00       4.25      17850  United Kingdom  
7   2010-12-01 08:28:00       1.85      17850  United Kingdom  
8   2010-12-01 08:28:00       1.85      17850  United Kingdom  
9   2010-12-01 08:34:00       1.69      13047  United Kingdom  
10  2010-12-01 08:34:00       2.10      13047  United Kingdom  
11  2010-12-01 08:34:00       2.10      13047  United Kingdom  
12  2010-12-01 08:34:00       3.75      13047  United Kingdom  
13  2010-12-01 08:34:00       1.65      13047  United Kingdom  
14  2010-12-01 08:34:00       4.25      13047  United Kingdom  
15  2010-12-01 08:34:00       4.95      13047  United Kingdom  
16  2010-12-01 08:34:00       9.95      13047  United Kingdom  
17  2010-12-01 08:34:00       5.95      13047  United Kingdom  
18  2010-12-01 08:34:00       5.95      13047  United Kingdom  
19  2010-12-01 08:34:00       7.95      13047  United Kingdom